### PR TITLE
Limit type name length

### LINKED ISSUE #13021: Put a limit on type name length
Index names already have a character limit of 255 characters or less. This should also be applied to types.

```http
PUT /my-index/type-0airn7cyol1jfddmrzax5xzw8nue8qkcetnhux2apco8rpi1iztwwq1vp65skvnolmal7ruoabkqtl6xurr0gwivkpmpu6kmmarjeugijzvfugv8ly0nbtleytjeruistrtmthmqyiuvrtwwzzt8gburbpiu1jpqprktjf7swesbjjenxnmjmypygwktgifyo6gigbv8wgkjained5pbikd5whybj22tas60rncmefbwb4xuampjvxbo3lpndlsoy8AFSADtsywxgkxawdfsl9maue1nfqysepuokgrybvhskhwn/1
{
  "field" : "value"
}
```

I cannot think of a legitimate use case for obscenely large type names and using them as keys is a mistake. If a user were to actually want this, then they can add a `not_analyzed` string field (perhaps `keyword` field in the future!) to support this behavior without abusing the type mappings.

### PR DESCRIPTION
**Pull request by jasontedor**
This commit will limit type name length to be at most 255 characters.

Closes #13021

### CODE REVIEW

**Comment by dadoonet** [comment]
Looks good.

May be tell people how much long it is for now? 

`mapping type name [" + mapper.type() + "] is " + mapper.type().length() + " characters long; limit is length 255`

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -261,6 +261,9 @@ private DocumentMapper merge(DocumentMapper mapper, boolean updateAllTypes) {
             if (mapper.type().length() == 0) {
                 throw new InvalidTypeNameException("mapping type name is empty");
             }
+            if (Version.indexCreated(indexSettings).onOrAfter(Version.V_2_0_0_beta1) && mapper.type().length() > 255) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: can you add a test for this? You can special the version created in the index settings ;)

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: [Added](https://github.com/elastic/elasticsearch/commit/e77f4e167c476b17427d8fe2f0f7fc55e6ed1f6f).

**Comment by s1monw** [comment]
LGTM

**Comment by jasontedor** [comment]
Good idea @dadoonet. I've [added](https://github.com/jasontedor/elasticsearch/commit/8bfabb14bc546895e8eb6515856c393267a40006) that.

**Comment by jasontedor** [comment]
Thanks @dadoonet and @s1monw for reviewing.

**Pull request merged**